### ISSUE TITLE
bless to a package name with 4 trailing colons doesn't produce the correct symbol table structure

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11245]
Migrated from [rt.perl.org#88134](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=88134) (status was 'resolved')

Searchable as RT88134$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



perlmod states&#8203;:

&nbsp;    The symbol table for a package happens to be stored in the hash of that
&nbsp;    name with two colons appended\.  The main symbol table's name is thus
&nbsp;    C\<%main&#8203;::>\, or C\<%&#8203;::> for short\.  Likewise the symbol table for the nested
&nbsp;    package mentioned earlier is named C\<%OUTER&#8203;::INNER&#8203;::>\.


The extreme corner case of bless with a package name ending with 4 \(or 3n \+ 4\)
colons doesn't produce the correct structure&#8203;:

$ cat bless\.pl 
\#\!/usr/bin/perl \-w
use strict;
use Data&#8203;::Dumper;
use Devel&#8203;::Peek;

$Data&#8203;::Dumper&#8203;::Indent = 1;
$Data&#8203;::Dumper&#8203;::Useqq = 1;

my $o = bless \[\]\, "foo&#8203;::&#8203;::";

print Dumper \\%foo&#8203;::;

Dump \*\{$foo&#8203;::\{"&#8203;::&#8203;:"\}\}\{HASH\}\->\{'&#8203;:'\};

warn $o;

\_\_END\_\_
$ \./perl \-Ilib bless\.pl 
$VAR1 = \{
&nbsp;  "&#8203;::&#8203;:" => \*foo&#8203;::&#8203;::&#8203;:
\};
SV = PVGV\(0xa51390\) at 0xa429a8
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(\)
&nbsp;  NAME = "&#8203;:"
&nbsp;  NAMELEN = 1
&nbsp;  GvSTASH = 0xa42978    "foo&#8203;::&#8203;:"
&nbsp;  GP = 0xa4cc60
&nbsp;    SV = 0x0
&nbsp;    REFCNT = 1
&nbsp;    IO = 0x0
&nbsp;    FORM = 0x0  
&nbsp;    AV = 0x0
&nbsp;    HV = 0xa42990
&nbsp;    CV = 0x0
&nbsp;    CVGEN = 0x0
&nbsp;    LINE = 9
&nbsp;    FILE = "bless\.pl"
&nbsp;    FLAGS = 0x0
&nbsp;    EGV = 0xa429a8      "&#8203;:"
foo&#8203;::&#8203;::=ARRAY\(0xa2fde8\) at bless\.pl line 15\.


For the case above\, Perl\_gv\_stashpvn\(\) passes "foo&#8203;::&#8203;::&#8203;::" to
Perl\_gv\_fetchpvn\_flags\(\) \(ie\, 6 colons total\)\. The parser in
Perl\_gv\_fetchpvn\_flags\(\) splits that as 'foo&#8203;::' '&#8203;::&#8203;:' with '&#8203;:' left over\.
If necessary it creates proper symbol tables for 'foo' and '&#8203;:'\. It returns
the HV for '&#8203;:'\, which Perl\_gv\_stashpvn\(\) "upgrades" to a symbol table by
giving it a name\.

Hence the final symbol table in the package doesn't end '&#8203;::'\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.14.0:

Configured by nick at Sat Apr  9 10:30:34 BST 2011.

Summary of my perl5 (revision 5 version 14 subversion 0) configuration:
  Derived from: d1ef1aa80747cb421af3e87bd1944a13a5fb2627
  Platform:
    osname=linux, osvers=2.6.35.4, archname=x86_64-linux
    uname='linux eris 2.6.35.4 #4 smp tue sep 21 09:54:22 bst 2010 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -DDEBUGGING -Doptimize=-Os -Uusethreads -Uuselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -Umad -de'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    RC0


@INC for perl 5.14.0:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0
    .


Environment for perl 5.14.0:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by 088225f\, as a result of fixing \#88132\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'resolved'
